### DISCUSSION TITLE
init prompts and existing/default values

### DISCUSSION THREAD

**Discussion by Lockszmith-GH** [twpayne/chezmoi#2546] [Q&A] [answered]
# My current state
Currently my `.chezmoi.toml.tmpl` file contians:
```
{{- $email := promptStringOnce . "email" "Email address" -}}

[data]
    pqb.email = {{ $email | quote }}
```

Today I edited the `.tmpl` file to include some default `diff` configs:
```
[diff]
    command = "icdiff"
    args = ["--line-numbers", "{{ "{{ .Destination }}" }}", "{{ "{{ .Target }}" }}"]
```

Now, running `chezmoi init` updates the existing configuration as intended, however...
It prompts for the email again.

# My question
Can the init assignment to $email grab the existing value and prompt only if the value is empty/non-existing?
#### bonus question
Can it populate default values to the prompt? So it will still prompt, but already populate the value with defaults (and populate that default value from the existing config)?

**Comment by twpayne** [comment]
This isn't working because `[data]` sets `pqb.email` whereas you search for `email`.

For default values you can use the [`default` template function from sprig](http://masterminds.github.io/sprig/defaults.html).

**Reply by Lockszmith-GH** [reply]
OK, so I probably stumbled upon a bug.

As you mentioned the following suppresses the prompt for email, if the email entry already exists.
```
{{- $email := promptStringOnce . "email" "Email address" -}}

[data]
    email = {{ $email | quote }}
```

This does not.
```
{{- $email := promptStringOnce . "pqb.email" "Email address" -}}

[data]
    pqb.email = {{ $email | quote }}
```

I'm guessing the `.` in this argument doesn't have a special meaning.

**Reply by twpayne** [reply]
No, the `.` does have a special meaning. You're using TOML and the `.` in `pqb.email` variable name is creating a nested object. Run `chezmoi data` and inspect the output.

**Reply by Lockszmith-GH** [reply]
That's what I expected from the `[data]` section.

I couldn't figure out how to access it from `promptStringOnce`. - That's where I meant 'no special meaning' - it seems it assumes it's part of the variable name and not a nesting separator.

Is there a different notation I need to use in the prompt directive for this to work?

**Reply by twpayne** [reply]
> Is there a different notation I need to use in the prompt directive for this to work?

No, `promptStringOnce` only knows about maps, not nested maps. You'll need to wrap the call to `promptStringOnce` in an extra layer of template logic to first check if `.pqb` exists, and then, if `.pqb` exists you can use `promptStringOnce .pqb "email" "Email address"`.

**Comment by twpayne** [comment, marked as answer]
With #2557, the `prompt*Once` template functions now take a path instead of a key, which allows you to use:

    {{- $email := promptStringOnce . "pqb.email" "Email address" -}}

**Reply by Lockszmith-GH** [reply]
Trigger warning: unsolicited complements and fandom.

I don't know if I expressed it enough, and I hope you enjoy the complement, since I'm being as sincere as one can be with the written word.

I enjoy using the tool you've shared with the world more than these words can express.

What makes this tool so awesome, though, is your interaction with us users here.
I personally have reached out a few times, and every time, you not only respond, you quickly address these conversations with meaningful action.

Here you already provided an elegant implemented solution before I even got to ask for it.

**Reply by Lockszmith-GH** [reply]
Perfect, works like a charm.
